proposed/activity-improvements.md

```diff
@@ -0,0 +1,327 @@
+if (activityListener.IsEnabled())
+{
+    Activity activity = null;
+    // Check if anyone cares about activity
+    if (activityListener.IsEnabled("Azure.Core.Http.Request"))
+    {
+        activity = new Activity("Azure.Core.Http.Request");
+        activity.AddTag..
+        activityListener.StartActivity(activity); // this does string concat and allocates every time
+    }
+
+    ...
+
+    if (activity != null)
+    {
+        activityListener.StopActivity(activity);
+    }
+}
+```
+
+Note in the pattern, the code trying to avoid creating the Activity object if there are no listeners interested to listen to such Activity creation. Also, the code later has to check if the activity object is created and then stop it.
+We are simplifying the pattern to something like:
+
+```C#
+    // StartNew can return null if no listener is enabled. Activity now is Disposable too which automatically stop the activity if it is created and started.
+    using (Activity activity = Activity.StartNew("Azure.Core.Http.Request")
+    {
+        activity?.AddTag..
+    }
+```
+
+- Last, we are introducing ActivityListener which can easily be used to listen to Activity objects for starting and stopping events. The new listener will help in the OT implementation as OT listen to the Activity events. It is now reasonable for a listener such as OpenTelemetry to observe Activity instrumentation from any code component in the app by the ActivityListener. Previously OpenTelemetry would have needed apriori knowledge of the strings that would be used for DiagnosticSource name and IsEnabled() callbacks for each Activity it wanted to receive callbacks for. When trying to create a new Activity object through the new introduced factory methods `StartNew`, the listener will have the opportunity to get a callback to decide if interested in having the Activity object get created or it can be sampled out to avoid the object creation if nobody else is listening or other listeners are not interested in such activity too. The listener will work with Activity's objects created using `new Activity(...)` so the listener will work with old written code using this pattern and can listen to the start and stop events for such Activity objects.
+
+Here is a couple of examples of how to create a listener and start it. 
+
+```C#
+        // The following pattern is useful for the Dependency Injection scenario to allow easily creating the listener with the reflection without the need to subclass and override the listener methods.
+        ActivityListener listener = ActivityListener.CreateAndStart(shouldCreateActivity: (name, parentContext, links) => true, onActivityStarted: (activity) => { ... }, onActivityStopped: (activity) => { ... });
+
+       // The following pattern is for apps or services need to write more code and add more properties to the listener as the DI pattern would not be good enough for them.
+        private class CustomListener : ActivityListener
```

Comment:
What can be achieved by subclassing that is not available using the `ActivityListener.CreateAndStart` ?

Reply:
Nothing more than if there is someone want to add more properties to the subclass for easier use.

Reply:
Should we cut down on public API surface and put the static method onto Activity class?

Reply:
I believe we can consider that. @noahfalk do you have any concern if we do that?

Reply:
The principle seems fine. We might want to add more optional virtual methods to the ActivityListener over time so I could imagine even if we don't make it public now we wind up adding it later. Exposing if/when we all believe it is justified seems like a fine plan though.

Here are some potential extra virtual methods I could imagine us wanting to add at some point:
a) CreateAndStart that takes a string rather than ActivityContext - this handles the hierarchial id case. I am proposing we add this one now.
b) IsEnabled(string activityName) - handles pre-filtering activities by name to use together with ActivitySource
c) CreateAndStart that takes a ReadOnlySpan<byte> utf8TraceParent, utf8TraceState - performance optimization for incoming http case that doesn't force parsing to occur prior to sampling.